### PR TITLE
Initial classes

### PR DESCRIPTION
**Pull request by samcogan21**
The initial code for our global optimization.

### CODE REVIEW

**Review by OfirSimhi1612 — COMMENTED** [review]
Looks very good overall, I've added few comments on the code let me know what you think. General thoughts I had:

1. Consider using `@classmethod` and `@staticmethod` in function declarations inside classes.
2. I would consider using types in function signatures (`def func(arg: str) -> List[int]:` for example), it helps understand how the functions should be used and prevents errors.
3. I think major part of the functions can be written as pure functions and not as class function. It can be useful to write them as pure function for future use as helpers, to logic that's not related to a specific class. It is also good to keep the classes with scoped logic and not make them "functions collections".
4. Adding lint + formatting is must at some point (in a different pr), it will prevent all the `unused` comments.

File: code/optimize.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    def run(self):
+        """
+        Function to run optimization
+        Returns:
+            best_solution (AdvancedMolecule)
+            best_fitness (float)
+            count (int): number of unsuccessful crossovers
+        """
+        for iteration in range(self.max_iter):
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/optimize.py:164]: Good to know! I'll make the change.

File: code/optimize.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            for i in range(self.population_size):
+                target = self.advanced_molecular_space.population[i]
+                mutant = self.mutate(i)
+                trial = self.crossover(target, mutant)
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/optimize.py:168]: That is on purpose. We used the logic of the differential evolution global optimization which uses a crossover function that outputs a trial vector. Essentially the crossover function goes over each value from both the original vector and the mutated vector (in this case the original molecule and mutated molecule) and randomly chooses one, which creates a random mix of both vectors- called the trial vector. In the case of a math function, it works well, but when dealing with molecules it creates invalid molecules 95+% of the time. That's why I didn't use it and compared the mutated molecule to the original one instead.

File: code/mol_space.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        """
+        Function to substitute atom with smart atom in order to increase chance of molecular validity.
+        It defines allowed substitutions based on atom type and valence
+        Argument:
+            self(Molecule): Molecule
+            molecule(Molecule): Molecule which is represented as fingerprint
+        Return:
+            mol(Molecule): Molecule which is represented as fingerprint
+        """
+        mol = Chem.RWMol(molecule.mol)
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/mol_space.py:81]: Good suggestion- I tried to do something like that by creating rules by which atoms could be replaced and what they can be replaced with. The tricky part of filtering out atoms that can't be replaced is that each molecule has different logic so it'd be really hard to come up with a strategy for that. Maybe we can brainstorm something together.

File: code/mol_space.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        if mol.GetNumAtoms() > 0:
+            atom_idx = random.randint(0, mol.GetNumAtoms() - 1)
+            atom = mol.GetAtomWithIdx(atom_idx)
+
+            substitutions = {
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/mol_space.py:86]: Cool, I'll implement this.

File: code/mol_space.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+            original_atom_symbol = atom.GetSymbol()
+            if original_atom_symbol in substitutions:
+                new_atom_symbol = random.choice(substitutions[original_atom_symbol])
+                new_atom = Chem.Atom(new_atom_symbol)
+                new_atom.SetFormalCharge(0)  # Reset formal charge
+                mol.ReplaceAtom(atom_idx, new_atom)
+
+        return self.sanitize_and_optimize_molecule(mol)
+
+    from rdkit import Chem
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/mol_space.py]: Fixed

File: code/mol_space.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
+from rdkit import Chem
+from rdkit.Chem import AllChem
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/mol_space.py]: Done

File: code/execute.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    # Calculate the total weight
+    total_weight = mw_weight + sascore_weight + ob_weight
+
+    # Compute the weighted average fitness score
+    fitness_score = (weighted_mw_score + weighted_sascore + weighted_ob_score) / total_weight
+
+    return fitness_score
+
+# Example usage
+initial_smiles = [
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/execute.py:45]: Done.

File: code/mol_space.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+            self(Molecule): Molecule
+            mol(Molecule): Molecule which is represented as fingerprint
+            k(int): Number of nearest neighbors
+        Return:
+            neighbors(list): List of nearest neighbors
+        """
+        distances = [(m, self.calculate_distance(mol, m)) for m in self.population if m != mol]
+        return sorted(distances, key=lambda x: x[1])[:k]
+
+    def mutate_molecule(self, molecule):
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/mol_space.py:46]: I like the idea of creating a seperate file for all the stand alone functions.

File: code/mol_space.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+                new_atom = Chem.Atom(new_atom_symbol)
+                new_atom.SetFormalCharge(0)  # Reset formal charge
+                mol.ReplaceAtom(atom_idx, new_atom)
+
+        return self.sanitize_and_optimize_molecule(mol)
+
+    from rdkit import Chem
+    from rdkit.Chem import AllChem
+
+    def find_connectable_atoms(self, mol):
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/mol_space.py:101]: Agreed, do you think that one file for all helper functions is good? Or multiple files? For example one helper file for each function that helps a certain class?

File: code/molecules.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        Returns:
+            ob_percentage (float): Oxygen bond percentage
+        """
+        formula = rdMolDescriptors.CalcMolFormula(self.mol)
+
+        # Regular expression to match elements and their counts
+        element_counts = re.findall(r'([A-Z][a-z]*)(\d*)', formula)
+
+        # Initialize counts for C, H, O, and M
+        x = 0  # Carbon count
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/molecules.py:52]: Good idea, will change.

File: code/optimize.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
+import numpy as np
+from rdkit import Chem
+from molecules import Molecule
+from mol_space import MolecularSpace
+from rdkit.Chem import BRICS
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/optimize.py:5]: You're right, used it in previous version but got rid of the function it was used in. Will delete.

File: code/optimize.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+            trial_mol = Chem.RWMol(target_mol)
+
+            for i in range(trial_mol.GetNumAtoms()):
+                if crossover_mask[i]:
+                    trial_mol.ReplaceAtom(i, mutant_mol.GetAtomWithIdx(i))
+
+            Chem.SanitizeMol(trial_mol)  # Try to sanitize
+            trial_smiles = Chem.MolToSmiles(trial_mol)
+            return Molecule(trial_smiles)
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (samcogan21) [reply code/optimize.py:132]: will correct.

File: code/optimize.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (OfirSimhi1612) [reply code/optimize.py:167]: So should we remove the whole row, or just change it to `self.crossover(target, mutant)` without defining new variable? (not urgent at all if this code is still in progress, feel free to resolve this conversation)

File: code/mol_space.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (OfirSimhi1612) [reply code/mol_space.py:101]: I think we should find files separation that makes sense, for example `mutations.py` can hold all mutations logic, etc.

File: code/mol_space.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (OfirSimhi1612) [reply code/mol_space.py:46]: great, we can figure out together the right separation for the files

**Pull request closed without merging**